ADR/nnnn-task-versioning.md

```diff
@@ -0,0 +1,427 @@
+- When the `0.x` version tag moves to a new revision, the older revision (previously
+  tagged as `0.x`) begins its expiry period.
+- *The same would likely apply to `1.x`, `2.x` etc. We don't know, no Konflux Task
+  has made it that far.*
+
+This mechanism is rather clunky and not very flexible. And, for reasons explained
+in [ADR NN](ADR/tbd.md), the mechanism will not work after the build-definitions
+decentralization.
+
+A more flexible solution would allow the policy author to allow/disallow specific
+version ranges. Such a solution cannot work if dozens of different versions of the
+same Task share the same version number.
+
+## Decision
+
+### Mark each meaningful change with a version update
+
+For Tasks at version `0.x`, you may continue to mark breaking changes with minor
+version updates and start marking non-breaking changes with patch version updates.
+Do aim to bring the Task to version `1.0`.
+
+For Tasks at version `1.0` or greater, follow [Semantic Versioning](https://semver.org/).
+
+This ADR leaves some freedom to each Task maintainer team to decide how to ensure
+changes get versioned. The *suggested* approach would be to version manually
+and set up CI in a way that encourages/enforces versioning. See Appendix 1:
+[The technicals of Task versioning](#the-technicals-of-task-versioning).
+
+### Don't require Task repositories to follow Tekton Catalog layout
+
+The [Tekton Catalog layout], which places each version of a Task in a separate
+`task/${task_name}/${version}/` directory, causes far too much overhead when
+updating Task versions. If we're going to require version updates for each meaningful
+change, we cannot ask Task maintainers to follow this layout. See Appendix 1:
+[The technicals of Task versioning](#the-technicals-of-task-versioning) for more details.
+
+### Keep a changelog
+
+Version numbers on their own are nice, but without an easy way to see what changed
+in each version, they're not that useful.
+
+Document the meaningful changes in each version in a CHANGELOG.md file. The changelog
+for a specific Task will be at `/task/${task_name}/CHANGELOG.md` in the repository.
+See Appendix 3: [CHANGELOG.md format](#changelogmd-format) for details about the
+file format.
+
+### MintMaker: don't assume each x.y version has a MIGRATION.md
+
+Today, MintMaker links MIGRATION.md documents in the PRs it sends to Konflux users
+(see the example [above](#users-get-opaque-updates)). This feature assumes that each
+`major.minor` version comes with a MIGRATION.md document, that this document is
+at `/task/${task_name}/${version}/MIGRATION.md` in the source repository and that
+the source repository is <https://github.com/konflux-ci/build-definitions>. The
+feature will not work after decentralization.
+
+MintMaker will stop trying to link MIGRATION.md files and will instead link the changelog.
+In case of breaking changes, the changelog should explain how the user can migrate.
```

Comment:
Is this proposing to do with automated remediation via the migration scripts?

Reply:
(But migration scripts will also still work)